src/medida/stats/sliding_window_sample.cc

```diff
@@ -0,0 +1,173 @@
+    const std::chrono::microseconds timeSlice_;
+    std::size_t admissionMask_;
+    std::default_random_engine rng_;
+    std::uniform_int_distribution<size_t> dist_;
+    std::deque<std::pair<double, Clock::time_point>> values_;
+};
+
+SlidingWindowSample::SlidingWindowSample(std::size_t windowSize,
+                                         std::chrono::seconds windowTime)
+    : impl_{new SlidingWindowSample::Impl{windowSize, windowTime}}
+{
+}
+
+SlidingWindowSample::~SlidingWindowSample()
+{
+}
+
+void
+SlidingWindowSample::Clear()
+{
+    impl_->Clear();
+}
+
+std::uint64_t
+SlidingWindowSample::size() const
+{
+    return impl_->size();
+}
+
+void
+SlidingWindowSample::Update(std::int64_t value)
+{
+    impl_->Update(value);
+}
+
+void
+SlidingWindowSample::Update(std::int64_t value, Clock::time_point timestamp)
+{
+    impl_->Update(value, timestamp);
+}
+
+Snapshot
+SlidingWindowSample::MakeSnapshot() const
+{
+    return impl_->MakeSnapshot();
+}
+
+// === Implementation ===
+
+SlidingWindowSample::Impl::Impl(std::size_t windowSize,
+                                std::chrono::seconds windowTime)
+    : windowSize_(windowSize)
+    , windowTime_(windowTime)
+    , timeSlice_(std::chrono::duration_cast<std::chrono::microseconds>(windowTime)/windowSize)
+    , admissionMask_(0)
+    , dist_(0, std::numeric_limits<size_t>::max())
+{
+    Clear();
+}
+
+SlidingWindowSample::Impl::~Impl()
+{
+}
+
+void
+SlidingWindowSample::Impl::Clear()
+{
+    std::lock_guard<std::mutex> lock{mutex_};
+    values_.clear();
+    admissionMask_ = 0;
+}
+
+std::uint64_t
+SlidingWindowSample::Impl::size()
+{
+    std::lock_guard<std::mutex> lock{mutex_};
+    return values_.size();
+}
+
+void
+SlidingWindowSample::Impl::Update(std::int64_t value)
+{
+    Update(value, Clock::now());
+}
+
+void
+SlidingWindowSample::Impl::Update(std::int64_t value,
+                                  Clock::time_point timestamp)
+{
+    if (!values_.empty())
+    {
+        // If we're in a new timeslice, reset the admission mask.
+        if (timestamp > values_.back().second + timeSlice_)
+        {
+            admissionMask_ = 0;
+        }
+
+        // If there's old data, trim it.
+        Clock::time_point expiryTime = timestamp - windowTime_;
+        while (values_.front().second < expiryTime)
+        {
+            values_.pop_front();
+        }
+    }
+
+    // When you add samples to the sliding window _slowly_ nothing goes wrong;
+    // when you add them too _quickly_ there's the possibility of losing rare
+    // events because they're overwritten before they get observed. To
+    // compensate for this, the sliding window maintains an "admission mask"
+    // that's reset on each timeslice, and used for exponentially-more-forceful
+    // stochastic rate limiting of each extra addition within a timeslice.
+    std::lock_guard<std::mutex> lock{mutex_};
+    if ((dist_(rng_) & admissionMask_) == 0)
+    {
+        values_.emplace_back(value, timestamp);
```

Comment:
actually, I am having a hard time thinking about the implications of this:
in "high volume" situation, we'd insert more than 1 item per timeslice, which means that we'd have in our buffer less than the target 5 minutes. For example, if the rate is 100 events per slice (2ms per item as a slice is ~200ms), we have an average of about 5 new samples per slice so the window will track roughly 1 minute worth of data.

I think a better strategy is to instead of adding elements, look at the last element inserted, and replace it.

Something like
```
    if ((dist_(rng_) & admissionMask_) == 0)
    {
        admissionMask_ <<= 1;
        admissionMask_ |= 1;
        // check if we've already inserted an item for the same timeSlice
        if (!values_.empty() && timestamp - timeSlice_ <= values_.back().second)
        {
            values_.back() = {value, timestamp};
        }
        else
        {
            values_.emplace_back(value, timestamp);
            if (values_.size() > windowSize_)
            {
                values_.pop_front();
            }
        }
    }
```

Reply:
Fair point, though the code as written doesn't work right -- it keeps extending the last-admitted timestamp forward with each replacement.

A version that keeps the old timestamp and only replaces the value works, though. Pushed that.